python_modules/dagster-graphql/dagster_graphql/implementation/asset_checks_loader.py

```diff
@@ -0,0 +1,121 @@
+from typing import Iterable, List, Mapping, Optional
+
+from dagster import (
+    _check as check,
+)
+from dagster._core.definitions.events import AssetKey
+from dagster._core.definitions.external_asset_graph import ExternalAssetGraph
+from dagster._core.host_representation.external_data import ExternalAssetCheck  # noqa: TCH002
+from dagster._core.storage.asset_check_execution_record import AssetCheckInstanceSupport
+from dagster._core.workspace.context import WorkspaceRequestContext
+from packaging import version
+
+from dagster_graphql.schema.asset_checks import (
+    AssetChecksOrErrorUnion,
+    GrapheneAssetCheck,
+    GrapheneAssetCheckCanExecuteIndividually,
+    GrapheneAssetCheckNeedsAgentUpgradeError,
+    GrapheneAssetCheckNeedsMigrationError,
+    GrapheneAssetCheckNeedsUserCodeUpgrade,
+    GrapheneAssetChecks,
+)
+
+
+class AssetChecksLoader:
+    """A batch loader that fetches asset check definitions for a set of asset keys."""
+
+    def __init__(self, context: WorkspaceRequestContext, asset_keys: Iterable[AssetKey]):
+        self._context = context
+        self._asset_keys = list(asset_keys)
+        self._checks: Optional[Mapping[AssetKey, AssetChecksOrErrorUnion]] = None
+
+    def _fetch_checks(self) -> Mapping[AssetKey, AssetChecksOrErrorUnion]:
+        instance = self._context.instance
+        asset_check_support = instance.get_asset_check_support()
+        if asset_check_support == AssetCheckInstanceSupport.NEEDS_MIGRATION:
+            return {
+                asset_key: GrapheneAssetCheckNeedsMigrationError(
+                    message="Asset checks require an instance migration. Run `dagster instance migrate`."
+                )
+                for asset_key in self._asset_keys
+            }
+        elif asset_check_support == AssetCheckInstanceSupport.NEEDS_AGENT_UPGRADE:
+            return {
+                asset_key: GrapheneAssetCheckNeedsAgentUpgradeError(
+                    "Asset checks require an agent upgrade to 1.5.0 or greater."
+                )
+                for asset_key in self._asset_keys
+            }
+        else:
+            check.invariant(
+                asset_check_support == AssetCheckInstanceSupport.SUPPORTED,
+                f"Unexpected asset check support status {asset_check_support}",
+            )
+
+        external_checks: Mapping[AssetKey, List[ExternalAssetCheck]] = {}
+        errors: Mapping[AssetKey, GrapheneAssetCheckNeedsUserCodeUpgrade] = {}
+
+        for location in self._context.code_locations:
+            for repository in location.get_repositories().values():
+                for external_check in (
```

Comment:
This triple for loop seems like it could get expensive on a big enough repository especially if we're only fetching 10 checks (thats the chunk size the UI uses) but there 1000+ checks. Is there a way we could have a mapping directly from asset key to external check and asset key to code location?

Reply:
Could you elaborate on the chunk size 10?

Reply:
Hmm I think what Marco is mentioning is that the UI asks for 10 asset nodes at a time when loading data for the asset graph. Eg:

```
assetNodes(assetKeys: [... 10 asset keys ]) {
    assetChecks {
      name
      canExecuteIndividually
      executionForLatestMaterialization {
        id
        runId
        status
        timestamp
        stepKey
        evaluation {
          severity
        }
      }
    }
}
```

Right now we don't have a `limit` on the number of asset checks per asset, though I think we've separately talked about limiting it to just 51, (and if we get 51, we'll say "50+ checks" and not try to report the exact count succeeded or failed on the graph)

Reply:
got it. Yeah we could improve this by storing an asset key -> check mapping, but we're getting away with the same loop for assets so I think we can do this for now https://github.com/dagster-io/dagster/blob/master/python_modules/dagster-graphql/dagster_graphql/implementation/fetch_assets.py#L128